### ISSUE TITLE
Any roadmap ahead?

### ISSUE THREAD

**Issue by deeplook** [amoffat/sh#450]
I like this module, but I see it hasn't been touched for a long time, many trivial PRs are piling up, questions are not answered. It would be nice to have a statement on further development maintenance or even some roadmap ahead from @amoffat. Or does anyone have pointers to more evolved forks of this?

**Comment by honnix** [comment]
+1 

I really like this project, so please don't let it stale.

**Comment by amoffat** [comment]
Thanks for your interest!  sh isn't dead, just dormant.  It is nearly feature complete, imho, so development cycles are quite a bit longer, but if you look at the commit history, I do find time to do releases (mostly around holidays).  One intentional side-effect of long development cycles is stability, and given the problem areas that sh intends to address, stability there is crucial.

As for questions, like most of you I am pretty busy, so I am relying on the supportive community to help out there.

All that said, if someone has great ideas for additional killer features that are not supported in sh, by all means, feel free to fork and implement them as an alternative project.

So to answer your initial question @deeplook, the roadmap ahead is primarily bug fixes and developing support for new versions of python.  Maybe async support is also within that.

**Comment by honnix** [comment]
I agree with you that `sh` is almost feature complete.

I also share the same concern as @deeplook about those trivial PRs and questions. They are contributions from the community but getting piling up. :(

I don't have any good idea. Maybe try to find another active maintainer so you will be able to share some workload?

**Comment by ssbarnea** [comment]
I am using sh in ansible/molecule and its current lack of maintenance worries me as it could easily become a liability. 

Can we maybe consider moving it under an umbrella organisation and assuring that there are at least 3 people that can review, merge and release? sh is great, lets assure it does not rot.

**Comment by amoffat** [comment]
@ssbarnea I appreciate your input, and I understand your concern.  If something critical pops up on sh, I will see it and address it.  In the "hit by a bus" scenario, fortunately there are many forks, and if you are concerned that a fix will not make it into master, I would encourage you to make a fork as well.  Most of the PRs that exist right now are related to documentation, typos, and logging output, so I don't foresee any deep breaking changes, but you're right, I can't know for sure. 

I'm less interested in a separate org, but I'm still open to it.  Are you volunteering to be a maintainer? 
 The long term goal of sh is to eventually get into the stdlib.  I don't know if that will happen because I don't have those kinds of connections, but that would be a preferable type of hand-off.

**Comment by ssbarnea** [comment]
@amoffat Thanks for the reply, loved it. I already have a list of project I maintian that keeps be more than busy so I should be realistical and I should not volunteer at this moment.  Also glad to hear that there is a "contingency plan".

Regarding stdlib.... that sounds nice but I do remember that a wise man said at some point that this is where good projects go to die. There was some truth in it.

**Comment by ssbarnea** [comment]
I don't know if that is irony but right after writing previous message I was hooked into making my first contribution https://github.com/amoffat/sh/pull/491 -- mainly because it was hit by the famous  `ErrorReturnCode_255` while working on some molecule changes.

**Comment by ssbarnea** [comment]
@amoffat can you please check https://github.com/amoffat/sh/pull/492 first? At least it brings CI to green, so we can properly test the other PRs.

Do you happen to be in a present in freenode  channel?

**Comment by amoffat** [comment]
@ssbarnea you should be good to proceed now with #491.  I'm not on IRC.  Are you looking for a way to collaborate on how to test #491?  If so, let's move the conversation to that issue.

**Comment by ssbarnea** [comment]
After reading https://github.com/amoffat/sh/pull/468/files I realise that I am probably better or removing use of sh from molecule. That patch mainly fixed the same stuff that I worked on yesterday and it was raised more than 6 months ago, totally ignored.

I also have zero interest in supporting python versions that reached EOL, there are already too many stable versions to support anyway. This is even worse when you realise that the project was not tested with py37 and py38. 

To summarise, my impression is that it would be easier to scrap the dependency instead of trying to fix it, there are just too many issues with it. Another examples of issues: not using standard python testing tools like tox, pytest and not even having linting. I did not make a decision yet.

**Comment by amoffat** [comment]
>I also have zero interest in supporting python versions that reached EOL

If you want to do that with your project, great, but how is this relevant to the goals of sh?

>This is even worse when you realize that the project was not tested with py37 and py38.

As currently advertised in the README.  Putting a stamp of approval on sh for a new version of Python can be more involved than you might think.  Some of the comments in the source code will explain why.

>Also, if merges are done into develop branch instead of master, github should be configured to have develop as default branch so incoming PRs are not going towards wrong branch. I am not a big fun of develop branch workflow and I know only one other major project using it

The default branch for this project is master, because I want users to be most likely to grab the most stable code.  If someone creates a PR against master instead of develop, it's not a big deal for me to change the base.  I prefer to cut releases from a develop branch, which leaves master free to receive emergency hotfixes.  This is a widely-used pattern in the git flow workflow.

>To summarize, my impression is that it would be easier to scrap the dependency instead of trying to fix it, there are just too many issues with it.

That's fine.

>Another examples of issues: not using standard python testing tools like tox, pytest and not even having linting.

There are reasons for many of these things, and again, if you read the source, or take the time to ask me, you will understand some of those reasons.  For example, tox is broken with 2.6, and I understand you don't care about EOL versions of Python, so you wouldn't care about being unable to test on 2.6, but I do.

>I did not make a decision yet.

Don't take this the wrong way, but whether or not you decide to use sh in your project, and keeping us in anticipation of your decision like this, is your personal choice and not relevant to this project.  Understand that sh underpins many pieces of infrastructure for many organizations, and **long-term stability** for those users is my highest priority for them.

If you choose to use sh, great, if you choose to use something else, great.

**Comment by ssbarnea** [comment]
Sorry for missing the reply notification. Shortly, I will be very happy to see https://github.com/amoffat/sh/pull/499/files merged. Once I know that sh is tested with all current versions of python I have enough trust to invest more in fixing other bugs.

I was worried that supporting py24 was more important than supporting py38. If that is not the case, I am more than happy.

**Comment by ssbarnea** [comment]
@amoffat Times passed and no progress is happening, neither PRs nor a new release since 2017. I keep seeing annoying warnings on projects where sh i used:
```
====================================================================== warnings summary ======================================================================
.tox/py27-ansible25-unit/lib/python3.7/site-packages/sh.py:424
  /Users/ssbarnea/c/os/molecule/.tox/py27-ansible25-unit/lib/python3.7/site-packages/sh.py:424: DeprecationWarning: invalid escape sequence \d
    rc_exc_regex = re.compile("(ErrorReturnCode|SignalException)_((\d+)|SIG[a-zA-Z]+)")
```
Yeah, that only a warning, but tomorrow it could be a blocker. Each time I see this is a reminder to me that `sh` package become a liability.

Can we please do something about it? I can understand that probably you do have other priorities but giving others permissions to review, merge and finally release new versions does not take too much time.

Lots of projects are affected by this kind of problem which unhanded will definitely kill them. For example [cookiecutter](https://pypi.org/project/cookiecutter/#history) was passing some rough times but lucky the maintainer did the right thing and added extra maintainers, [unblocking](https://github.com/cookiecutter/cookiecutter/pulse) the project.

**Comment by amoffat** [comment]
Please see https://github.com/amoffat/sh/issues/512
